lib/logstash/inputs/kafka.rb

```diff
@@ -249,6 +252,14 @@ class LogStash::Inputs::Kafka < LogStash::Inputs::Base
 
   attr_reader :metadata_mode
 
+  def initialize(params = {})
+    unless params.key?('codec')
+      params['codec'] = params.key?('schema_registry_url') ? 'json' : 'plain'
+    end
+
+    super(params)
```

Comment:
I have thought about it a lot, and this is certainly the best place to perform the injection. We get fully-contextualized codecs as a result.

This has the minor side-effects of (1) putting `codec` in the `original_params` when it wasn't specified and (2) mutating the hash that was sent. 

I'm not sure it's worth the effort, but the following doesn't have those side-effects.

```suggestion
    return super if params.key?('codec')
    
    params_with_codec = params.merge('codec' => (params.key?('schema_registry_url') ? 'json' : 'plain')
    super(params_with_codec)
    @original_params.delete('codec') # we were never here...
```

Reply:
both the `super` (from [`config_init`](https://github.com/elastic/logstash/blob/v7.15.2/logstash-core/lib/logstash/config/mixin.rb#L94)) and super's super ([`Plugin#initialize`](https://github.com/elastic/logstash/blob/v7.15.2/logstash-core/lib/logstash/plugin.rb#L79)) change the input `params`.

the work here was inspired by existing behavior -> fill in the `params['codec']` default before `config_init`

this is not ideal but the contract is broken already in core in multiple places, at this point I would be more concerned about breaking smt that relies on the existing core behavior of passing in `params` and having all defaults filled in in the same Hash instance.